use xfs_repair to check and repair xfs filesystem in method formatAndMount

27149chen opened the pull request:
This PR tries to fix the issue mentioned in kubernetes/utils issues/125, changes are as following:
1. move the general check and repair method (using fsck) to an individual function.
2. add a new function "checkAndRepairXfsFilesystem" to check and repair xfs filesystem.
3. add a new function "analyzeXfsRepairOutput" to analyze the output of xfs_repair output because the exit status is always 0.
4. in method formatAndMount, add a switch to use different functions for different filesystems.
5. fix two xfs test case and add two more.

Which are not included in this PR:
1. A clean log on a file system is required for xfs_repair to operate. If the file system was not cleanly unmounted, it should be mounted and unmounted prior to using xfs_repair

References:
1. https://linux.die.net/man/8/xfs_repair
2. https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/storage_administration_guide/fsck-fs-specific#fsck-xfs

Commit:
use xfs_repair to check and repair xfs filesystem

Signed-off-by: Lou <luogj@cn.ibm.com>

cheneyzhao commented on mount/mount_linux.go:
```diff
@@ -271,21 +367,16 @@ func (mounter *SafeFormatAndMount) formatAndMount(source string, target string,
 
 	if !readOnly {
 		// Run fsck on the disk to fix repairable issues, only do this for volumes requested as rw.
```
This comment is not accurate.  Suggest "Run fsck or xfs_repair on the disk to fix repairable issues based on file system type ..."

cheneyzhao commented on mount/mount_linux.go:
```diff
@@ -257,6 +258,101 @@ func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
 	return SearchMountPoints(realpath, procMountInfoPath)
 }
 
+// checkAndRepairFileSystem is a general method to check and repair filesystem
```
general to default?

cheneyzhao commented on mount/mount_linux.go:
```diff
@@ -257,6 +258,101 @@ func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
+			if err != nil {
+				return fmt.Errorf("'xfs_repair' found errors on device %s but could not correct them: %v", source, err)
+			} else {
+				return nil
```
Can have a log here, such as klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)

cheneyzhao commented on mount/mount_linux.go:
```diff
@@ -257,6 +258,101 @@ func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
+		case isExitError && ee.ExitStatus() == fsckErrorsUncorrected:
+			return fmt.Errorf("'fsck' found errors on device %s but could not correct them: %s", source, string(out))
+		case isExitError && ee.ExitStatus() > fsckErrorsUncorrected:
+			klog.Infof("`fsck` error %s", string(out))
```
Will Errorf be better here？

cheneyzhao reviewed (approved):
/lgtm

xqlixqli commented:
/lgtm

cheneyzhao reviewed (approved):
/lgtm

gnufied committed:
Add myself as review for mount directory

gnufied committed:
Return typed error when Mount Fails

Mount can fail for a variety of reasons and caller might want to know
why mount failed. Using untyped string based error does not provide
enough granularity to make that verification.

k8s-ci-robot committed:
Merge pull request #128 from gnufied/add-as-reviewer

Add myself as reviewer for mount directory

k8s-ci-robot committed:
Merge pull request #129 from gnufied/add-error-type-for-mount-failure

Return typed error when Mount Fails

Commit:
update after review

Signed-off-by: Lou <luogj@cn.ibm.com>
